compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -297,7 +297,9 @@ fn implicit_negative<'cx, 'tcx>(
             predicate: p,
         })
         .chain(obligations)
-        .find(|o| !selcx.predicate_may_hold_fatal(o));
+        .find(|o| {
+            !selcx.evaluate_root_obligation(o).map(|result| result.may_apply()).unwrap_or_default()
```

Comment:
one tiny last comment:

can you do `map_or(true, ..)` instead of `unwrap_or_default`? And also, this needs a comment that explains that overflow is treated as if the obligation may apply.